File: editors/vscode/src/server/platform.ts
import * as os from "os";
import { BINARY_NAME, BINARY_ROOT_DIR } from "./constants";

// Types

export type OS = "windows" | "macos" | "linux";
export type Arch = "x86_64" | "aarch64";

// Constructor partials

function constructOs(): OS {
	switch (os.platform()) {
		case "win32":
			return "windows";
		case "darwin":
			return "macos";
		case "linux":
			return "linux";
		default:
			throw new Error(`Unsupported OS: ${os.platform()}`);
	}
}

function constructArch(): Arch {
	switch (os.arch()) {
		case "x64":
			return "x86_64";
		case "arm64":
			return "aarch64";
		default:
			throw new Error(`Unsupported architecture: ${os.arch()}`);
	}
}

function constructExeSuffix(): string {
	return os.platform() === "win32" ? ".exe" : "";
}

// Strings class

export class PlatformDescriptor {
	private readonly os: OS;
	private readonly arch: Arch;
	private readonly exeSuffix: string;

	constructor() {
		this.os = constructOs();
		this.arch = constructArch();
		this.exeSuffix = constructExeSuffix();
	}

	public releaseAssetName(version: string): string {
		const cleanVersion = version.startsWith("v")
			? version.slice(1)
			: version;
		return `${BINARY_NAME}-${cleanVersion}-${this.os}-${this.arch}.zip`;
	}

	public serverBinaryRoot(): string {
		return BINARY_ROOT_DIR;
	}

	public serverBinaryDir(version: string): string {
		return `${BINARY_NAME}-${version}`;
	}

	public serverBinaryPath(): string {
		return `${BINARY_NAME}${this.exeSuffix}`;
	}

	public isUnix(): boolean {
		return this.os === "linux" || this.os === "macos";
	}
}

File: editors/zed/src/platform.rs
use zed_extension_api::{self as zed};

use crate::constants::{BINARY_NAME, BINARY_ROOT_DIR};

#[derive(Debug, Clone, Copy)]
pub struct PlatformDescriptor {
    os: &'static str,
    arch: &'static str,
    exe_suffix: &'static str,
}

impl PlatformDescriptor {
    pub fn current() -> Self {
        let (platform, arch) = zed::current_platform();

        let os = match platform {
            zed::Os::Windows => "windows",
            zed::Os::Linux => "linux",
            zed::Os::Mac => "macos",
        };

        let arch = match arch {
            zed::Architecture::Aarch64 => "aarch64",
            _ => "x86_64",
        };

        let exe_suffix = match platform {
            zed::Os::Windows => ".exe",
            zed::Os::Linux | zed::Os::Mac => "",
        };

        Self {
            os,
            arch,
            exe_suffix,
        }
    }

    pub fn release_asset_name(&self, version: &str) -> String {
        format!(
            "{BINARY_NAME}-{}-{}-{}.zip",
            version.trim_start_matches('v'),
            self.os,
            self.arch
        )
    }

    pub fn server_binary_root(&self) -> String {
        BINARY_ROOT_DIR.to_string()
    }

    pub fn server_binary_dir(&self, version: &str) -> String {
        format!("{BINARY_NAME}-{version}")
    }

    pub fn server_binary_path(&self) -> String {
        format!("{BINARY_NAME}{}", self.exe_suffix)
    }
}
